service/src/event/service.rs

```diff
@@ -129,23 +130,77 @@ impl CeramicEventService {
         )?)
     }
 
-    pub(crate) async fn insert_events(
-        &self,
-        items: &[recon::ReconItem<EventId>],
-        source: DeliverableRequirement,
-    ) -> Result<InsertResult> {
-        let mut to_insert = Vec::new();
-        let mut invalid = Vec::new();
+    async fn validate_signed_events(
+        events: Vec<EventInsertable>,
+    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        // TODO: IMPLEMENT THIS
+        Ok((events, Vec::new()))
+    }
+
+    async fn validate_time_events(
+        events: Vec<EventInsertable>,
+    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        // TODO: IMPLEMENT THIS
+        Ok((events, Vec::new()))
+    }
+
+    pub(crate) async fn validate_events(
+        items: &[ReconItem<EventId>],
+    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        let mut parsed_events = Vec::with_capacity(items.len());
+        let mut invalid_events = Vec::new();
         for event in items {
             match Self::parse_discovered_event(event).await {
-                Ok(insertable) => to_insert.push(insertable),
-                Err(err) => invalid.push(InvalidItem::InvalidFormat {
+                Ok(insertable) => parsed_events.push(insertable),
+                Err(err) => invalid_events.push(InvalidItem::InvalidFormat {
                     key: event.key.clone(),
                     reason: err.to_string(),
                 }),
             }
         }
 
+        // Group events by their type
+        let mut valid_events = Vec::with_capacity(parsed_events.len());
+        let mut signed_events = Vec::with_capacity(parsed_events.len());
```

Comment:
I'm over-allocating here, since events will either be signed or time, each of these will definitely have less than `parsed_events.len()`.  But I figured over-allocating a little bit is better than having multiple allocations as the vector grows, and keeping these allocations a somewhat consistent size might make it easier for the allocator to be more efficient with re-using the memory.